### PR TITLE
Extending the Real Time Data Module

### PR DESCRIPTION
**Pull request by omerBrowsi**
Real time data module according to - 
https://github.com/prebid/Prebid.js/issues/4610

### CODE REVIEW

**Comment by karimMourra** [comment]
@omerBrowsi an integration test failed because `array.find` is not supported. This is fixed in https://github.com/prebid/Prebid.js/pull/5537/commits/34b9f9b96b33dcd4a9b1de304392e8c5402af1da

File: modules/rtdModule/index.js [hunk 1]
```diff
@@ -97,61 +172,74 @@ export function init(config) {
...
 
+/**
+ * call each sub module init function by config order
+ * if no init function / init return failure / module not configured - remove it from submodules list
+ */
+export function initSubModules(subModules) {
+  let subModulesByOrder = [];
+  _dataProviders.forEach(provider => {
+    const sm = subModules.find(s => s.name === provider.name);
+    const initResponse = sm && sm.init && sm.init(provider, gdprDataHandler.getConsentData(), uspDataHandler.getConsentData()) !== 'failure';
```
Comment (jsnellbaker) [inline modules/rtdModule/index.js:183]: To clarify on the GDPR and USP data, are we sure that in the sequence of things that this data would be available by the time the submodule's `init` functions are called?  

This `initSubModules` function is loaded as part of the overall rtdModule's `init()`, which in turn is loaded as the file is loaded.

But the consent information is obtained once the `requestBids` call chain starts, which should be after this code here has already executed.  So could there be a scenario here where these consent params in the `sm.init` wouldn't have the real data?  Or am I misreading something?

Comment (omerBrowsi) [reply modules/rtdModule/index.js]: Yes, you are correct
I added a hook so it will run before 'makeBidRequests', if you think of a better solution please let me know

File: modules/rtdModule/index.js [hunk 2]
```diff
@@ -23,14 +23,56 @@
...
- * @summary sub module name
- * @name ModuleConfig#name
- * @type {string}
+ * @function
+ * @summary init sub module
+ * @name RtdSubmodule#init
+ * @param {Object} config
+ * @param {Object} gdpr settings
+ * @param {Object} usp settings
+ * @return {string|boolean} "failure" to remove sub module
```
[the remark this answers is not in the document]
Comment (omerBrowsi) [reply modules/rtdModule/index.js]: it's not. changed to false

File: modules/rtdModule/index.js [hunk 3]
```diff
@@ -97,61 +172,74 @@ export function init(config) {
...
 /**
  * get data from sub module
  * @param {AdUnit[]} adUnits received from auction
  * @param {function} callback callback function on data received
  */
-function getProviderData(adUnits, callback) {
-  const callbackExpected = subModules.length;
-  let dataReceived = [];
+export function getProviderData(adUnits, callback) {
+  const mustWaitSubModulesLength = subModules.filter(sm => sm.config && sm.config.waitForIt).length;
```
[the remark this answers is not in the document]
Comment (omerBrowsi) [reply modules/rtdModule/index.js]: all submodules are async
the wait for is actually a way to say what submodules are less important - meaning we prefer not to delay the auction any more for them

I did change the names a bit, let me know what you think

File: modules/rtdModule/index.js [hunk 4]
```diff
@@ -40,25 +82,53 @@
...
- * @type {number}
+ * @summary name
+ * @name ModuleConfig#name
+ * @type {string}
+ */
+
+/**
+ * @property
+ * @summary delay auction for this sub module
+ * @name ModuleConfig#waitForIt
```
[the remark this answers is not in the document]
Comment (omerBrowsi) [reply modules/rtdModule/index.js:120]: all are async, so it might be confusing. please see comment on the second converstion

File: modules/rtdModule/index.js [hunk 5]
```diff
@@ -97,61 +172,74 @@ export function init(config) {
...
  * @param {AdUnit[]} adUnits received from auction
  * @param {function} callback callback function on data received
  */
-function getProviderData(adUnits, callback) {
-  const callbackExpected = subModules.length;
-  let dataReceived = [];
+export function getProviderData(adUnits, callback) {
+  const mustWaitSubModulesLength = subModules.filter(sm => sm.config && sm.config.waitForIt).length;
+  let callbackExpected = mustWaitSubModulesLength || subModules.length;
+  const mustHaveModules = mustWaitSubModulesLength > 0;
```
[the remark this answers is not in the document]
Comment (omerBrowsi) [reply modules/rtdModule/index.js]: changed the names a bit and added comments on the function, hope it makes things more clear

File: modules/rtdModule/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (karimMourra) [reply modules/rtdModule/index.js]: I understand now thank you so much for the clarification! I think "topPriority" would be a good term; it might be easier for future readers

**Comment by karimMourra** [comment]
Hi @omerBrowsi and @jsnellbaker is there an ETA for when this branch will be merged ? It's currently a blocker for https://github.com/prebid/Prebid.js/pull/5537

**Comment by Asafsham** [comment]
Hey @robertrmartinez, would you be able to get to this PR? tools PMC really wants to push it asap.

**Comment by bretg** [comment]
We already have two reviewers, so taking Bobby off as a required reviewer. @Asafsham - we're going to a document at some point soon: a 'how to write an RTD module' guide similar to https://docs.prebid.org/dev-docs/integrate-with-the-prebid-analytics-api.html

But I'm not going to call that a blocker. So if @pm-harshad-mane is willing to push his unit test splitting request to a future PR, this can be merged.

File: src/targeting.js [hunk 6]
```diff
@@ -325,7 +325,10 @@ export function newTargeting(auctionManager) {
       Object.keys(targetingConfig).filter(customSlotMatching ? customSlotMatching(slot) : isAdUnitCodeMatchingSlot(slot))
         .forEach(targetId =>
           Object.keys(targetingConfig[targetId]).forEach(key => {
-            let valueArr = targetingConfig[targetId][key].split(',');
+            let valueArr = targetingConfig[targetId][key];
+            if (typeof valueArr === 'string') {
+              valueArr = valueArr.split(',');
+            }
             valueArr = (valueArr.length > 1) ? [valueArr] : valueArr;
```
Comment (bretg) [inline src/targeting.js:332]: I don't know why this change is being made, but seems like it could cause an error if it gets to this line and valueArr is still a string, then valueArr.length is going to fail?

**Comment by Asafsham** [comment]
@bretg @pm-harshad-mane We are working on the Docs, should be ready be Monday.
